### PR TITLE
First PoC of a RSS notifier

### PR DESCRIPTION
**Pull request by LucasVerneyDGE**
Here is a very first example of what the RSS output discussed in https://github.com/ambanum/CGUs/issues/119 could look like.

So far, a single RSS feed is generated, bundling all the notifications. This is a strict RSS analogous to the email notifications provided by SendInBlue.

Below is an example of the generated RSS feed (running only with a single service, from an empty database):

```xml
<?xml version="1.0" encoding="utf-8"?>
<rss version="2.0" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:content="http://purl.org/rss/1.0/modules/content/">
    <channel>
        <title>Latest CGUs document updates</title>
        <link>https://github.com/ambanum/CGUs-versions/commit/</link>
        <description>Latest CGUs document updates from https://github.com/ambanum/CGUs.</description>
        <lastBuildDate>Fri, 11 Sep 2020 12:48:36 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>https://github.com/jpmonette/feed</generator>
        <item>
            <title><![CDATA[Google Play just updated their Terms of Service]]></title>
            <link>https://github.com/ambanum/CGUs-versions/commit/0dad2ab</link>
            <guid>https://github.com/ambanum/CGUs-versions/commit/0dad2ab</guid>
            <pubDate>Fri, 11 Sep 2020 12:48:34 GMT</pubDate>
            <content:encoded><![CDATA[<p>We recorded a new version of Google Play's Terms of Service.</p><p><a href="https://github.com/ambanum/CGUs-versions/commit/0dad2ab"><button style="background-color: #007bff;padding: .375rem .75rem;border: 1px solid transparent;color: #fff;border-radius: .25rem;line-height: 1.5;">View the changes</button></a></p><p>If the button does not work, copy the following address into your browser address bar: https://github.com/ambanum/CGUs-versions/commit/0dad2ab.</p>]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[Google Play just updated their Developer Terms]]></title>
            <link>https://github.com/ambanum/CGUs-versions/commit/d438795</link>
            <guid>https://github.com/ambanum/CGUs-versions/commit/d438795</guid>
            <pubDate>Fri, 11 Sep 2020 12:48:34 GMT</pubDate>
            <description><![CDATA[We recorded a new version of Google Play's Terms of Service. View the changes at https://github.com/ambanum/CGUs-versions/commit/d438795.]]></description>
            <content:encoded><![CDATA[<p>We recorded a new version of Google Play's Terms of Service.</p><p><a href="https://github.com/ambanum/CGUs-versions/commit/d438795"><button style="background-color: #007bff;padding: .375rem .75rem;border: 1px solid transparent;color: #fff;border-radius: .25rem;line-height: 1.5;">View the changes</button></a></p><p>If the button does not work, copy the following address into your browser address bar: https://github.com/ambanum/CGUs-versions/commit/d438795.</p>]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[Google Play just updated their Review Guidelines]]></title>
            <link>https://github.com/ambanum/CGUs-versions/commit/73cb0bc</link>
            <guid>https://github.com/ambanum/CGUs-versions/commit/73cb0bc</guid>
            <pubDate>Fri, 11 Sep 2020 12:48:36 GMT</pubDate>
            <description><![CDATA[We recorded a new version of Google Play's Terms of Service. View the changes at https://github.com/ambanum/CGUs-versions/commit/73cb0bc.]]></description>
            <content:encoded><![CDATA[<p>We recorded a new version of Google Play's Terms of Service.</p><p><a href="https://github.com/ambanum/CGUs-versions/commit/73cb0bc"><button style="background-color: #007bff;padding: .375rem .75rem;border: 1px solid transparent;color: #fff;border-radius: .25rem;line-height: 1.5;">View the changes</button></a></p><p>If the button does not work, copy the following address into your browser address bar: https://github.com/ambanum/CGUs-versions/commit/73cb0bc.</p>]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[Google Play just updated their Brand Guidelines]]></title>
            <link>https://github.com/ambanum/CGUs-versions/commit/8b77028</link>
            <guid>https://github.com/ambanum/CGUs-versions/commit/8b77028</guid>
            <pubDate>Fri, 11 Sep 2020 12:48:36 GMT</pubDate>
            <description><![CDATA[We recorded a new version of Google Play's Terms of Service. View the changes at https://github.com/ambanum/CGUs-versions/commit/8b77028.]]></description>
            <content:encoded><![CDATA[<p>We recorded a new version of Google Play's Terms of Service.</p><p><a href="https://github.com/ambanum/CGUs-versions/commit/8b77028"><button style="background-color: #007bff;padding: .375rem .75rem;border: 1px solid transparent;color: #fff;border-radius: .25rem;line-height: 1.5;">View the changes</button></a></p><p>If the button does not work, copy the following address into your browser address bar: https://github.com/ambanum/CGUs-versions/commit/8b77028.</p>]]></content:encoded>
        </item>
    </channel>
</rss>
```

Best

### CODE REVIEW

File: src/app/history/recorder.js [hunk 1]
```diff
@@ -30,14 +31,13 @@ export default class Recorder {
   }
 
   async save({ serviceId, documentType, content, fileExtension }) {
-    const directory = `${this.path}/${serviceId}`;
+    const filePath = this.getPathFor(serviceId, documentType, fileExtension);
```
Comment (LucasVerneyDGE) [inline src/app/history/recorder.js:34]: I went with tweaking a bit the `Recorder` to be able to reuse all of its logic even for storing files at the root of the versions repository (hence not belonging to a particular service provider).

As a result, I noticed a logical duplication in the code here, computing the path twice. I edited this part to actually first compute the full path and then ensuring required folder exists through `dirname`.

File: src/app/history/recorder.js [hunk 2]
```diff
@@ -84,6 +84,9 @@ export default class Recorder {
   }
 
   getPathFor(serviceId, documentType, fileExtension) {
+    if (serviceId === null) {
```
Comment (LucasVerneyDGE) [inline src/app/history/recorder.js:87]: When no service provider is specified, documents are stored at the root of the repo.

**Comment by Ndpnt** [comment]
I don't know if you already knew this feature but Github provides RSS feed on a repo https://github.com/ambanum/CGUs-versions/commits.atom .
In our case it won't allow us to notify users only for subset of services but I found it an interesting feature.

**Comment by Ndpnt** [comment]
It also work on a subfolder history. So if you want to follow, for example, only Google you can use this feed url:
https://github.com/ambanum/CGUs-versions/commits/master/Google.atom

**Comment by Ndpnt** [comment]
If you agree, I think we could replace the code of this PR by a documentation on how to get feed url for only one service (`https://github.com/ambanum/CGUs-versions/commits/master/<serviceId>.atom`) or even for a specific document (`https://github.com/ambanum/CGUs-versions/commits/master/<serviceId>/<documentType>.md.atom`)

**Comment by LucasVerneyDGE** [comment]
> 
> 
> It also work on a subfolder history. So if you want to follow, for example, only Google you can use this feed url:
> https://github.com/ambanum/CGUs-versions/commits/master/Google.atom
> 
> So it could do the job for our use case, what do you think?

I knew about the global RSS feed for a directory, but I definitely missed that Github was also offering them for subfolders (quite a hidden feature!). This could definitely do the job.

What about adding for example a generic "README.md" into each service folder (https://github.com/ambanum/CGUs-versions/tree/master/Google for example) linking to the correct RSS feed and such info.

My only concern with the Github approach is to have a too tight dependency on Github, with issues such as https://github.com/ambanum/CGUs-versions/issues/5 in mind. Otherwise, let's go for it.

**Comment by Ndpnt** [comment]
> What about adding for example a generic "README.md" into each service folder (https://github.com/ambanum/CGUs-versions/tree/master/Google for example) linking to the correct RSS feed and such info.

As we consider CGUs-versions as a database only we decided to remove all documentation from its repository (also because commits not related to document versions caused us some troubles) and to centralize documentation in this repository. 
So I'm not in favor of adding files in the versions' "database", but we definitively need a clear documentation to allow user to easily subscribe to document updates.
I think a section in the documentation and also a generated list of all available URLs feeds could do the job, what do you think? and you @MattiSG ?

> My only concern with the Github approach is to have a too tight dependency on Github, with issues such as [ambanum/CGUs-versions#5](https://github.com/ambanum/CGUs-versions/issues/5) in mind. Otherwise, let's go for it.

I totally understand your concern and we aim to become independent from Github! 
For now we use it because it allow us to provide features at a minimal cost and we try to demonstrate the value of the product. We currently think that it is more valuable to work on the core of the tool than to develop a UI or a feature service which could be provided in another way.

**Comment by MattiSG** [comment]
> It also work on a subfolder history.

OMG that's awesome 😱

> we definitively need a clear documentation to allow user to easily subscribe to document updates

Yes please, let's add that to the user's manual.

> a generated list of all available URLs feeds

Let's just document how to get the RSS URL from the service name. When we want to be more user-friendly we'll have to create a full UI anyway 😉 I'm not sure of the added value of a list of RSS feeds at the moment.

> My only concern with the Github approach is to have a too tight dependency on Github

I agree! But I'm way more in favor of delivering cheap'n'easy for now. We first need to get traction. Once there is an ecosystem consuming RSS feeds it'll be much easier to define the most appropriate file format, necessary metadata etc etc. It's not like we were going with full vendor lock-in: RSS is standard, it would mostly be a matter of changing the base URL to switch to another server. On the other hand, going this way saves us the hassle of hosting a public server, generating RSS etc etc… and focus on value proposition definition 🙂

**Comment by LucasVerneyDGE** [comment]
Definitely fine for me to go with the Github approach and just document it. There is no need to go further with this RSS PoC then, on generating per-service RSS feeds.

Should we still review this existing codebase to generate a global RSS feed, as an example of alternative notifier mechanism, if it's working fine?

**Comment by Ndpnt** [comment]
We will keep this branch and we will look at this PoC as a base for discussion when the need of an alternative notifier mechanism will come back, thanks for that! :)

**Pull request closed without merging**